fitbenchmarking/fitbenchmarking
# Added support for python 3.12

## Description of work

Fixes #1291 
Fixes #1255
Fixes #1254

- Updated `base_controller.py` to use import alias for `numpy`. This is consistent with how `numpy` is used across the repository.
- Removed the `numpy<1.24` restriction to make fitbenchmarking installable on the latest (stable) version of python `3.12`. 
- The latest version of the `pylint` and `flake8` packages will be installed on the GitHub runners. The code has been updated to meet the latest formatting standards. Linting tests now run on the latest version of python.
   - `pylint==2.9.5` -> `pylint==3.2.6`
   - `flake8==3.9.2` -> `flake8==7.1.1`
-  `actions/upload-artifact@v3` -> `actions/upload-artifact@v4`. [v3](https://github.com/actions/upload-artifact) will be depreciated in a few months.
-  `actions/download-artifact@v3` -> `actions/download-artifact@v4`. [v3](https://github.com/actions/download-artifact) will be depreciated in a few months.
-  `actions/checkout@v3` -> `actions/checkout@v4`. [v4](https://github.com/actions/checkout) is the latest version.
- Updated the GitHub runners to test installation of `fitbenchmarking` on python `3.8` and `3.12`. The default `linux` and `windows` test will be run on each of the 2 versions of python.
- Bumped the minimum supported version of [python](https://devguide.python.org/versions/) to `3.8.1+`. However, this will need to be bumped in a few months when `3.8` reaches the end of its life.
- Removed the [Derivative Free Optimiser - Gauss Newton](https://github.com/numericalalgorithmsgroup/dfogn) `"dfogn"` minimiser options for `"dfo"` software. This is installable in python 3.12 but will raise an error during runtime due to depreciated `numpy` methods. [Derivative Free Optimiser for Least Squares](https://github.com/numericalalgorithmsgroup/dfols) `"dfols"` which is the latest and more flexible version of `"dfogn"` is still available through `fitbenchmarking` options.
- Links in the documentation for `DFO-LS` were updated as the previous links are out of date.
- Updated `nightly.yml` to use the latest version of python instead of python `3.9`
- Updated `release.yml` to use the latest version of python instead of python `3.8`
- Updated docs for `"gofit"` software.




## Testing Instructions



1. run `make html` in the `docs` folder. Verify the docs are updated with the relevant information.

## Checklist
